### PR TITLE
Makes the bell silent

### LINKED ISSUE #1162: Repeatedly outputting `BEL` doesn't produce any sound
### Contour Terminal version

Contour Terminal Emulator 0.3.13-master-362a2868

### Installer source

Github: source code cloned

### Operating System

Fedora 38

### Architecture

x86-64

### Other Software

_No response_

### Steps to reproduce

Run `for i in $(seq 20);do printf '\a'; sleep 1; done` in bash 

### Expected Behavior

Contour should ring bell every one second.

### Actual Behavior

Contour doesn't produce any sound.

### Additional notes

Other ways this can be repro'd is by doing `printf'\a' ` a bunch of time in shell, after a while no sound is produced for BEL character.

### PR DESCRIPTION
**Pull request by Utkarsh-khambra**
Adds a config option `bell` under profile to configure BEL behaviour and Fixes #1162 ,  #1163

### CODE REVIEW

File: src/contour/Config.h [hunk 1]
```diff
@@ -275,8 +276,19 @@ struct Config
...
+        assert(prof);
+        return *prof;
+    }
+
+    TerminalProfile const& profile() const noexcept
+    {
+        auto const* prof = profile(defaultProfileName);
+        assert(prof);
+        return *prof;
+    }
```
Comment (Utkarsh-khambra) [inline src/contour/Config.h:291]: This is silence gcc error warning about possible nullptr deference.

**Review by Yaraslaut — CHANGES_REQUESTED** [review]

**Comment by Yaraslaut** [comment]
Also, we need to update documentation here  [profiles.md](https://github.com/contour-terminal/contour/blob/master/docs/configuration/profiles.md)

File: docs/configuration/profiles.md [hunk 2]
```diff
@@ -80,6 +80,16 @@ profiles:
 
 
 
+### `bell`
+configuration option determines whether or not BEL character '\a' or 0x07 produces audible sound or not.
```
Comment (christianparpart) [inline docs/configuration/profiles.md:84]:
```suggestion
Configuration option determines whether or not `BEL` character `\a` or `0x07` produces audible sound or not.
```

File: docs/configuration/profiles.md [hunk 3]
```diff
@@ -80,6 +80,16 @@ profiles:
...
+``` yaml
+profiles:
+  profile_name:
+    bell: true
+```
+
+
+
 ### `wm_class`
 configuration option defines the class part of the WM_CLASS property of the terminal window. The WM_CLASS property is a standard X11 property used to identify and classify windows by their class and instance.
```
Comment (christianparpart) [inline docs/configuration/profiles.md:94]:
```suggestion
Configuration option defines the class part of the `WM_CLASS` property of the terminal window. The `WM_CLASS` property is a standard X11 property used to identify and classify windows by their class and instance.
```

**Review by christianparpart — CHANGES_REQUESTED** [review]
LGTM, only very minors, that might make sense to apply. What do you think?

**Comment by Utkarsh-khambra** [comment]
I read the suggestion making bell to be `off` , `default` or `path`  to make more configurable. Working on that.

**Review by Yaraslaut — CHANGES_REQUESTED** [review]
Looks like not all wav files are good for a bell sound, i just downloaded random one from https://mixkit.co/free-sound-effects/bell/  and it didn't work :(

**Comment by Utkarsh-khambra** [comment]
@Yaraslaut can link a specific one? I tried a bunch of 'em from here already and they worked.

File: src/contour/ui.template/Terminal.qml [hunk 4]
```diff
@@ -59,9 +59,16 @@ ContourTerminal
...
+        id: bellAudioOutput
+        objectName: "BellAudioOutput"
+    }
+
+    MediaPlayer {
         id: bellSoundEffect
-        source: vtWidget.bellSoundSource
+        objectName: "Bell"
+        source: "qrc:/contour/bell.wav"
+        audioOutput: bellAudioOutput
```
Comment (christianparpart) [inline src/contour/ui.template/Terminal.qml:71]: Only the `.qml.in` files were supposed to be git-tracked in the `ui.template` folder. You can basically drop the changes here and it'll have no affect.

In a future/other PR we should actually get rid of these. :)

File: src/contour/contour.yml [hunk 5]
```diff
@@ -176,6 +176,9 @@ profiles:
         # whether or not to put the window into maximized mode.
         maximized: false
 
+        # Determines if BEL character sound is "off" or "default" or is given by file at "path"
```
Comment (christianparpart) [inline src/contour/contour.yml:179]: I think we can make this a multi-line comment, specifically describing each value.

File: src/contour/ContourGuiApp.cpp [hunk 6]
```diff
@@ -389,6 +389,16 @@ int ContourGuiApp::terminalGuiAction()
...
     newWindow();
 
+    if (auto const& bell = config().profile().bell; bell == "off")
+    {
+        qmlEngine_->rootObjects().first()->findChild<QObject*>("BellAudioOutput")->setProperty("muted", true);
+    }
+    else if (bell != "default")
+    {
+        QUrl const path(bell.c_str());
+        qmlEngine_->rootObjects().first()->findChild<QObject*>("Bell")->setProperty("source", path);
```
Comment (christianparpart) [inline src/contour/ContourGuiApp.cpp:399]: Can `qmlEngine_->rootObjects().first()->findChild<QObject*>("Bell")` return `nullptr`? If so, we should dereference it only if it's assured to be non-null.

File: src/contour/ContourGuiApp.cpp [hunk 7]
```diff
@@ -389,6 +389,16 @@ int ContourGuiApp::terminalGuiAction()
     // Spawn initial window.
     newWindow();
 
+    if (auto const& bell = config().profile().bell; bell == "off")
+    {
+        qmlEngine_->rootObjects().first()->findChild<QObject*>("BellAudioOutput")->setProperty("muted", true);
```
Comment (christianparpart) [inline src/contour/ContourGuiApp.cpp:394]: same here (like below, sorry for the inverse comments) :)

File: src/contour/Config.h [hunk 8]
```diff
@@ -42,6 +42,8 @@
 #include <unordered_map>
 #include <variant>
 
+#include "crispy/assert.h"
```
Comment (christianparpart) [inline src/contour/Config.h:45]:
```suggestion
#include <crispy/assert.h>
```

**Review by christianparpart — CHANGES_REQUESTED** [review]
Sorry, I should have used the review functionality from the start. :)

**Review by christianparpart — APPROVED** [review]
Many thanks

**Pull request merged**